### PR TITLE
Differentiate liveness and readiness probes for router pods

### PR DESCRIPTION
**Pull request by JacobTanenbaum**
Create upstream commit that allows for multiple groups of checks to be associated with health checking. Using the multiple groupings differentiate the liveness and readiness probes for the Haproxy router

Bug [1550007](https://bugzilla.redhat.com/show_bug.cgi?id=1550007)

### CODE REVIEW

**Comment by knobunc** [comment]
@smarterclayton is this a reasonable way to approach the problem?

**Comment by knobunc** [comment]
@openshift/networking PTAL

File: pkg/oc/admin/router/router.go [hunk 1]
```diff
@@ -436,7 +436,7 @@ func generateSecretsConfig(cfg *RouterConfig, namespace string, defaultCert []by
 	return secrets, volumes, mounts, nil
 }
 
-func generateProbeConfigForRouter(cfg *RouterConfig, ports []kapi.ContainerPort) *kapi.Probe {
+func generateProbeConfigForRouter(Path string, cfg *RouterConfig, ports []kapi.ContainerPort) *kapi.Probe {
```
Comment (rajatchopra) [inline pkg/oc/admin/router/router.go]: camelCase for 'Path'

**Comment by pravisankar** [comment]
@rajatchopra I don't think there will be part 2 here.
https://github.com/openshift/origin/pull/19009/files#diff-43ed12581474d0d4e9bd5b12bdcf6417R40
creates additional '/livez' endpoint (eg: http://localhost:1936/livez) which always returns success and this is used for liveness probe. Success from this endpoint means router is up but not ready until /healthz returns success.

File: pkg/oc/admin/router/router.go [hunk 2]
```diff
@@ -447,7 +447,7 @@ func generateProbeConfigForRouter(cfg *RouterConfig, ports []kapi.ContainerPort)
 		}
 
 		probe.Handler.HTTPGet = &kapi.HTTPGetAction{
-			Path: "/healthz",
+			Path: path,
 			Port: intstr.IntOrString{
 				Type:   intstr.Int,
 				IntVal: int32(healthzPort),
```
Comment (rajatchopra) [inline pkg/oc/admin/router/router.go]: Even though a path is being passed (which is either livez or healthz), the port is always healthz?

Comment (pravisankar) [reply pkg/oc/admin/router/router.go]: Yes, both /livez and /healthz paths are using StatsPort
https://github.com/JacobTanenbaum/origin/blob/1f93335b3f25944c202f7f05e3e52201d07ff972/pkg/cmd/infra/router/template.go#L206 and
https://github.com/JacobTanenbaum/origin/blob/1f93335b3f25944c202f7f05e3e52201d07ff972/pkg/oc/admin/router/router.go#L446

File: pkg/oc/admin/router/router.go [hunk 3]
```diff
@@ -436,21 +435,21 @@ func generateSecretsConfig(cfg *RouterConfig, namespace string, defaultCert []by
...
+func generateProbeConfigForRouter(path string, cfg *RouterConfig, ports []kapi.ContainerPort) *kapi.Probe {
 	var probe *kapi.Probe
 
 	if cfg.Type == "haproxy-router" {
 		probe = &kapi.Probe{}
-		healthzPort := defaultHealthzPort
+		probePort := defaultStatsPort
 		if cfg.StatsPort > 0 {
-			healthzPort = cfg.StatsPort
+			probePort = cfg.StatsPort
```
Comment (JacobTanenbaum) [inline pkg/oc/admin/router/router.go]: @rajatchopra @pravisankar does this make it clearer? removing the reference to the healthzPort in favour of a more generic term

Comment (pravisankar) [reply pkg/oc/admin/router/router.go]: +1

File: pkg/oc/admin/router/router.go [hunk 4]
```diff
@@ -466,15 +465,15 @@ func generateProbeConfigForRouter(cfg *RouterConfig, ports []kapi.ContainerPort)
 }
 
 func generateLivenessProbeConfig(cfg *RouterConfig, ports []kapi.ContainerPort) *kapi.Probe {
-	probe := generateProbeConfigForRouter(cfg, ports)
+	probe := generateProbeConfigForRouter("/livez", cfg, ports)
```
Comment (smarterclayton) [inline pkg/oc/admin/router/router.go]: This should be `/healthz/ready` which is our standard readiness check path

Comment (smarterclayton) [reply pkg/oc/admin/router/router.go]: `/healthz` is liveness, `/healthz/ready` is readiness.

File: pkg/router/metrics/metrics.go [hunk 5]
```diff
@@ -37,6 +37,10 @@ type Listener struct {
 func (l Listener) handler() http.Handler {
 	mux := http.NewServeMux()
 	healthz.InstallHandler(mux, l.Checks...)
+	mux.Handle("/livez", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (smarterclayton) [inline pkg/router/metrics/metrics.go]: What are you actually trying to check for?

**Comment by smarterclayton** [comment]
This should be following our standard conventions, but you cannot change the meaning of the existing health endpoint without breaking backwards compatibility.

Can you describe what the goal is here - have the endpoint available for load balancers *prior* to the router actually being loaded?  Generally if you want that you just use the service field `spec.publishNotReadyAddresses`, which doesn't need any of these changes.

**Comment by knobunc** [comment]
@smarterclayton the router added ROUTER_BIND_PORTS_AFTER_SYNC so that _external_ load balancers could have useful health checks on port 80 or 443 for the router to know when it was online.  Before we added ROUTER_BIND_PORTS_AFTER_SYNC the router would bind those ports immediately and potentially serve HTTP 503 statuses for valid routes because the route had not been loaded yet.  So, we added ROUTER_BIND_PORTS_AFTER_SYNC so that it would not bind 80/443 until a full sync had happened, BUT would bind 1936 so that the liveness probes worked (otherwise the router gets killed).

After the refactor to make the router controller return status directly, if you set ROUTER_BIND_PORTS_AFTER_SYNC then you start failing liveness probes.  @JacobTanenbaum found out that this was due to the liveness check done by the router controller contacting the haproxies it controlled to see if they were up, and if not, returning false.  So, when ROUTER_BIND_PORTS_AFTER_SYNC is set, the haproxy doesn't bind 80 and 443 so the delegated liveness check fails.  And the pod gets terminated.

So, the goal here is to have an endpoint indicating that the router controller is live, but that it is not ready yet... but we have to work with what we have at the moment:

- /healthz -- Returns only when the backends are ready (changed in 3.7)
- /healthz/backend-http -- Returns only when haproxy is up (added in 3.7)

Before 3.7 /healthz was a liveness check

Can we call this a bugfix and add /healthz/ready and move the multiplexer that is on /healthz there?  (Should we also support /healthz/ready/backend-http since that's what the multiplexer will set up)?

Then /healtz can go back to returning true as soon as the router controller becomes active.

**Comment by smarterclayton** [comment]
The liveness check is to prevent a crashed haproxy router from staying dead
due to a route controller bug.  The primary purpose is to keep the pod
running.  The fundamental behavior of that check didn't change (pre 3.7, if
that failed haproxy was dead, and post 3.7, if that failed haproxy is
dead), but the secondary effect you're referencing did.

No external load balancer should be using /healthz to determine whether to
put the router in rotation - it should be using /healthz/ready or something
equivalent.  If we want to health check the router controller we should add
`/healthz/controller` or similar.

When we move to service load balancer for router on cloud providers, the
readiness check for the router service should be /healthz/backend-http (if
you only want to be in rotation if haproxy is listening), and set the
service to preserveUnreadyEndpoints (if you want to be in rotation
regardless of readiness)

**Comment by knobunc** [comment]
@smarterclayton The problem is that the external load balancers are looking at 80 and 443 directly to decide whether to put it in rotation.  BUT the liveness check defined for the pod is on /healthz and that is checking that the haproxy is live too before returning 200.  So when we don't bind to 80 the liveness probes fail.  What would you suggest we do to make the liveness check work?

I see two options for this

1. Document that this option is deprecated and only works when the old stats are in use and then create and document proper readiness and liveness checks
2. Work out some way to make the option work with the current router so we can provide a liveness check

**Comment by smarterclayton** [comment]
Sorry for not responding faster, this is on my list to respond to.

**Comment by JacobTanenbaum** [comment]
@smarterclayton PTAL

File: pkg/oc/admin/router/router.go [hunk 6]
```diff
@@ -466,15 +465,15 @@ func generateProbeConfigForRouter(cfg *RouterConfig, ports []kapi.ContainerPort)
 }
 
 func generateLivenessProbeConfig(cfg *RouterConfig, ports []kapi.ContainerPort) *kapi.Probe {
-	probe := generateProbeConfigForRouter(cfg, ports)
+	probe := generateProbeConfigForRouter("/healthz/ready", cfg, ports)
```
Comment (smarterclayton) [inline pkg/oc/admin/router/router.go]: There's no way we can use these names without it being utterly confusing.  Your readiness probe is using our canonical liveness probe path, and your liveness probe is using our canonical readiness path.

Change the path you added to the router to `/healthz/controller` (the health of the controller) and that mitigates it.

However, this is still not the correct default configuration.  By default we should be using `/healthz` for liveness.  It's only when you want the "don't bind until loaded" behavior that these should flip.  Unless we switched to "don't bind until loaded" by default in the router and I missed it?

**Comment by knobunc** [comment]
Ok, so:

- Add `/healthz/controller` so it always returns true
- Add `/healthz/ready` so that it returns true when the haproxy is up and synchronized
- Change the behavior of `/healthz` so that when `BindPortsAfterSync` is set it returns true, otherwise it behaves as it does today by checking the endpoints

Change `oc adm router` so that readiness always goes to `/healthz/ready`

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go [hunk 7]
```diff
@@ -66,16 +66,25 @@ func NamedCheck(name string, check func(r *http.Request) error) HealthzChecker {
 // exactly one call to InstallHandler. Calling InstallHandler more
 // than once for the same mux will result in a panic.
 func InstallHandler(mux mux, checks ...HealthzChecker) {
+	InstallPathHandler(mux, "/healthz", checks...)
```
Comment (pravisankar) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go]: Do we need to change vendor code, can't we do this stuff in openshift code?

Comment (pravisankar) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go]: AFAIK changing vendor code needs 'UPSTREAM: <carry>' commit in this case.

Comment (JacobTanenbaum) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go]: @pravisankar I did post an upstream PR to occupancy this [kubernetes/PR63716](https://github.com/kubernetes/kubernetes/pull/63716). The commit that includes the kubernetes code is tagged UPSTREAM:63716, should I add this tag to the PR title?

**Comment by knobunc** [comment]
@JacobTanenbaum Can you fix your description to capture the current behavior please.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go (hunk 7 above)
Comment (pravisankar) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go]: Separate UPSTREAM:63716 commit is good, no need to add upstream tag to the PR title.

**Comment by knobunc** [comment]
@deads2k can you approve the upstream commit to vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver please.  (Since https://github.com/kubernetes/kubernetes/pull/63716 landed upstream)

Thanks

**Comment by knobunc** [comment]
@deads2k -- can you approve this since the upstream PR has merged.

File: pkg/router/metrics/health.go [hunk 8]
```diff
@@ -35,6 +36,28 @@ func HTTPBackendAvailable(u *url.URL) healthz.HealthzChecker {
 	})
 }
 
+// HasSynced returns a healthz check that verifies the router has been synced at least
+// once.
+func HasSynced(router **templateplugin.TemplatePlugin) healthz.HealthzChecker {
```
Comment (smarterclayton) [inline pkg/router/metrics/health.go]: This sort of construct is a violation of our style guides.  You should be passing an interface here that exposes the correct check method.  Double pointers should never be used.  Nothing in metrics should be aware of template plugin at all.

File: pkg/router/metrics/health.go [hunk 9]
```diff
@@ -35,6 +36,28 @@ func HTTPBackendAvailable(u *url.URL) healthz.HealthzChecker {
...
+	return healthz.NamedCheck("has-synced", func(r *http.Request) error {
+		if router != nil {
+			if (*router).Router.SyncedAtLeastOnce() == true {
```
Comment (smarterclayton) [inline pkg/router/metrics/health.go]: This construct is not appropriate.  It should always be `if booleancondition {`

File: pkg/router/metrics/health.go [hunk 10]
```diff
@@ -35,6 +36,28 @@ func HTTPBackendAvailable(u *url.URL) healthz.HealthzChecker {
...
+				return nil
+			} else {
```
Comment (smarterclayton) [inline pkg/router/metrics/health.go]: When you return early, elide the else, as per our style guide.

File: pkg/router/metrics/health.go [hunk 11]
```diff
@@ -35,6 +36,28 @@ func HTTPBackendAvailable(u *url.URL) healthz.HealthzChecker {
...
+				return fmt.Errorf("Router not synced")
```
Comment (smarterclayton) [inline pkg/router/metrics/health.go]: Errors should always be lower case, as per the style guide.

**Pull request merged**